src/hooks/users.ts

```diff
@@ -1,63 +1,88 @@
-import { useMemo } from "react";
-import { isLoaded } from "react-redux-firebase";
-
-import { User } from "types/User";
-import { ValidStoreAsKeys } from "types/Firestore";
-
-import { usersSelector, usersByIdSelector } from "utils/selectors";
+import { worldUsersSelector, usersByIdSelector } from "utils/selectors";
 import { WithId } from "utils/id";
 
-import { useSelector } from "hooks/useSelector";
-import { useUserLastSeenThreshold } from "hooks/useUserLastSeenThreshold";
-import { useVenueId } from "hooks/useVenueId";
-import { useFirestoreConnect } from "hooks/useFirestoreConnect";
+import { useSelector } from "./useSelector";
+import { useUserLastSeenThreshold } from "./useUserLastSeenThreshold";
+import { useConnectCurrentVenueNG } from "./useConnectCurrentVenueNG";
+import { useVenueId } from "./useVenueId";
+import { useFirestoreConnect, isLoaded } from "./useFirestoreConnect";
 
-const useConnectVenueUsers = () => {
+import { User } from "types/User";
+import { useMemo } from "react";
+
+export const useConnectWorldUsers = () => {
   const venueId = useVenueId();
 
-  // @debt refactor this + related code so as not to rely on using a shadowed 'storeAs' key
-  //   this should be something like `storeAs: "currentVenueUsers"` or similar
-  useFirestoreConnect(
-    venueId
-      ? {
-          collection: "users",
-          where: ["enteredVenueIds", "array-contains", venueId],
-          storeAs: "users" as ValidStoreAsKeys, // @debt super hacky, but we're consciously subverting our helper protections
-        }
-      : undefined
+  const { isCurrentVenueLoaded, currentVenue } = useConnectCurrentVenueNG(
+    venueId!
   );
-};
 
-export const useVenueUsers = (): readonly WithId<User>[] => {
-  useConnectVenueUsers();
+  useFirestoreConnect(() => {
+    if (!isCurrentVenueLoaded || !currentVenue) return [];
+
+    const relatedLocationIds = [currentVenue.id];
 
-  return useSelector(usersSelector) ?? [];
+    if (currentVenue.parentId) {
+      relatedLocationIds.push(currentVenue.parentId);
+    }
+
+    return [
+      {
+        collection: "users",
+        where: ["enteredVenueIds", "array-contains-any", relatedLocationIds],
+        storeAs: "worldUsers",
+      },
+    ];
```

Comment:
You are right, but as per our discussion on one of daily meetings, I think we have come to a conclusion that we can sacrifice the edge case when we have venue => parentVenue => parentParentVenue. I might be wrong though. This logic will also be changed with https://github.com/sparkletown/internal-sparkle-issues/issues/127 implemented

Reply:
That's not my understanding of that meeting. My understanding was that it's crucial that we support the full multi-level hierarchy for counting to be correct.

I believe one of my past suggestions for how we could resolve this is that we could 'walk' the full path the first time the page is loaded, and then store those results, which we can then use for this query. Another alternative would be to 'pre-compute' this chain at the time a venue is created/updated (in the admin panel), and store this data in firestore somewhere. Or achieve that same functionality, but have it triggered from a firestore 'event triggered function (https://firebase.google.com/docs/functions/firestore-events) rather than handling it from the frontend.